### ISSUE TITLE
The --dry-run option should return a non-zero exit code

### ISSUE THREAD

**Issue by Soullivaneuh** [rectorphp/rector#800]
Or if it's not possible, maybe add a `--exit-code` option as you can see on the `git diff` command.

It is for CI usage.

Related to https://github.com/rectorphp/rector/issues/799#issuecomment-442900482

**Comment by TomasVotruba** [comment]
:+1:

**Comment by TomasVotruba** [comment]
Is this standard in CLI tools? 

`--dry-run` should return non-zero on activity, it just doesn't perform any change

**Comment by Soullivaneuh** [comment]
PHPUnit always return a non-zero exit code, but it's not comparable because it's a testing tool.

PHP-CS-Fixer **does** return a non-zero exit code.

But I don't thing any standard exist about this. This is also why I proposed an alternative with the `--exit-code` option.

**Comment by TomasVotruba** [comment]
What do you mean by `--exit-code`?

**Comment by TomasVotruba** [comment]
I think supressing exit code is big wtf. This should be part of CI setup, like `allowed_failure`

**Comment by Soullivaneuh** [comment]
`--exit-code` may be used to add an non-zero exit code if rector did any change on the code, with dry option or not.

I didn't follow you for the second comment, sorry. Could you please elaborate?

**Comment by TomasVotruba** [comment]
> --exit-code may be used to add an non-zero exit code if rector did any change on the code, with dry option or not.

Could you show me some example? I don't understand. Now Rector returns 1 on fail, 0 on success. That should remain the same.

> I didn't follow you for the second comment, sorry. Could you please elaborate?

In `travis.yml`: https://github.com/rectorphp/rector/blob/8696ef2c6284c5cc188aaa2ff3d5c04c4f2f6653/.travis.yml#L15-L17

**Comment by Soullivaneuh** [comment]
> Could you show me some example? I don't understand.

Of course:

```bash
$ git diff
diff --git a/Makefile b/Makefile
index 8de963d03..0d142f8ea 100644
--- a/Makefile
+++ b/Makefile
@@ -59,6 +59,9 @@ cs-fix:
 
 rector:
        # @see https://github.com/rectorphp/rector/issues/799
+       ${RECTOR} php70 -vvv
+       ${RECTOR} php71
+       ${RECTOR} php72
        ${RECTOR} phpunit70
        ${RECTOR} symfony40
        ${RECTOR} symfony41
$ echo $?
0
$ git diff --exit-code
diff --git a/Makefile b/Makefile
index 8de963d03..0d142f8ea 100644
--- a/Makefile
+++ b/Makefile
@@ -59,6 +59,9 @@ cs-fix:
 
 rector:
 	# @see https://github.com/rectorphp/rector/issues/799
+	${RECTOR} php70 -vvv
+	${RECTOR} php71
+	${RECTOR} php72
 	${RECTOR} phpunit70
 	${RECTOR} symfony40
 	${RECTOR} symfony41
$ echo $?
1
```

> Now Rector returns `1` on code change

Are you sure about that? 

```
sullivan@1f214b7e994e:/code$ vendor/bin/rector process src --level symfony42
Rector v0.3.20
Config file: /code/rector.yml

 850/850 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

1 Changed file
==============

 * /code/src/AppBundle/DependencyInjection/Configuration.php

                                                                                                                        
 [OK] Rector is done!                                                                                                   
                                                                                                                        

sullivan@1f214b7e994e:/code$ echo $?
0
```

But to me, it's normal. No crash, so 0. This is why I suggest a return code on `--dry-run` or add a `--exit-code` option. :+1:

**Comment by TomasVotruba** [comment]
Thanks :+1: 

> Are you sure about that?

I meant required code change.

I look at the docs and it look like it inverts the exit code: https://git-scm.com/docs/git-diff#git-diff---exit-code

**Comment by Soullivaneuh** [comment]
> it look like it inverts the exit code

Yes, it clearly designed for automation.

**Comment by TomasVotruba** [comment]
So it fails when it should be successful?

**Comment by Soullivaneuh** [comment]
In pure technical representation, yes. :wink:

**Comment by Soullivaneuh** [comment]
And no. Because in case of crash, the exit code would not be 0, even with the `--exit-code` option.

**Comment by TomasVotruba** [comment]
All right. Will you implement it?

**Comment by Soullivaneuh** [comment]
To be clear, with `--exit-code`, `git diff` will:

- Return `0` if no diff
- Return `1` if diff
- Return something else different from `0` if it crash

Is that more clear?

**Comment by Soullivaneuh** [comment]
I might take few minutes to take a look, yes. But do you prefer the `--dry-run` option modification of the `--exit-code` addition? I would vote for the first possibility.

**Comment by TomasVotruba** [comment]
> Is that more clear?

Yes, thank you.

**Comment by TomasVotruba** [comment]
Lets go with `--dry-run` option modification for now and see how community takes it.
We can always add `--exit-code` later.

**Comment by Soullivaneuh** [comment]
I did a PR.

I think #799 should be re-open because it's two complete different features.

**Comment by TomasVotruba** [comment]
> I did a PR.

:+1: 

> I think #799 should be re-open because it's two complete different features.

Please comment the PR under the PR. This is confusing.

**Comment by Soullivaneuh** [comment]
I'm talking about issue #799, not the PR i just opened.

If you merge the PR I created, #800 will be closed and #799 never resolved.

**Comment by TomasVotruba** [comment]
> I'm talking about issue #799, not the PR i just opened.

I know, comment under #799 for #799
